File: WinPaletter/Extras/ScreenSaver_Editor.vb
Imports System.ComponentModel

Public Class ScreenSaver_Editor

    Private Proc As Process

    Private Sub ScreenSaver_Editor_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        LoadLanguage
        ApplyStyle(Me)
        XenonButton12.Image = MainFrm.XenonButton20.Image.Resize(16, 16)
        pnl_preview.DoubleBuffer
        ApplyFromCP(My.CP)
    End Sub


    Sub ApplyFromCP(CP As CP)
        With CP.ScreenSaver
            ScrSvrEnabled.Checked = .Enabled
            XenonTextBox1.Text = .File
            XenonTrackbar5.Value = .TimeOut
            XenonCheckBox1.Checked = .IsSecure
        End With

    End Sub

    Sub ApplyToCP(CP As CP)
        With CP.ScreenSaver
            .Enabled = ScrSvrEnabled.Checked
            .File = XenonTextBox1.Text
            .TimeOut = XenonTrackbar5.Value
            .IsSecure = XenonCheckBox1.Checked
        End With
    End Sub

    Private Sub XenonButton11_Click(sender As Object, e As EventArgs) Handles XenonButton11.Click
        If OpenFileDialog1.ShowDialog = DialogResult.OK Then
            Dim CPx As New CP(CP.CP_Type.File, OpenFileDialog1.FileName)
            ApplyFromCP(CPx)
            CPx.Dispose()
        End If
    End Sub

    Private Sub XenonButton9_Click(sender As Object, e As EventArgs) Handles XenonButton9.Click
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyFromCP(CPx)
        CPx.Dispose()
    End Sub

    Private Sub XenonButton12_Click(sender As Object, e As EventArgs) Handles XenonButton12.Click
        Using _Def As CP = CP_Defaults.From(My.PreviewStyle)
            ApplyFromCP(_Def)
        End Using
    End Sub

    Private Sub XenonButton8_Click(sender As Object, e As EventArgs) Handles XenonButton8.Click
        ApplyToCP(My.CP)
        Close()
    End Sub

    Private Sub XenonButton10_Click(sender As Object, e As EventArgs) Handles XenonButton10.Click
        Cursor = Cursors.WaitCursor
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyToCP(CPx)
        ApplyToCP(My.CP)
        CPx.ScreenSaver.Apply()
        CPx.Dispose()
        Cursor = Cursors.Default
    End Sub

    Private Sub XenonButton7_Click(sender As Object, e As EventArgs) Handles XenonButton7.Click
        Close()
    End Sub

    Private Sub ScrSvrEnabled_CheckedChanged(sender As Object, e As EventArgs) Handles ScrSvrEnabled.CheckedChanged
        checker_img.Image = If(sender.Checked, My.Resources.checker_enabled, My.Resources.checker_disabled)
    End Sub

    Private Sub XenonButton5_Click(sender As Object, e As EventArgs) Handles XenonButton5.Click
        If IO.File.Exists(XenonTextBox1.Text) AndAlso IO.Path.GetExtension(XenonTextBox1.Text).ToUpper = ".SCR" Then
            If Proc IsNot Nothing AndAlso Not Proc.HasExited Then Proc.Kill()
            Proc = Process.GetProcessById(Shell("""" & XenonTextBox1.Text & """" & " /p " & pnl_preview.Handle.ToInt32))
        End If
    End Sub


    Private Sub XenonButton1_Click(sender As Object, e As EventArgs) Handles XenonButton1.Click
        If OpenFileDialog2.ShowDialog = DialogResult.OK Then
            XenonTextBox1.Text = OpenFileDialog2.FileName
        End If
    End Sub

    Private Sub XenonTextBox1_TextChanged(sender As Object, e As EventArgs) Handles XenonTextBox1.TextChanged
        If IO.File.Exists(XenonTextBox1.Text) AndAlso IO.Path.GetExtension(XenonTextBox1.Text).ToUpper = ".SCR" Then
            If Proc IsNot Nothing AndAlso Not Proc.HasExited Then Proc.Kill()
            Proc = Process.GetProcessById(Shell("""" & XenonTextBox1.Text & """" & " /p " & pnl_preview.Handle.ToInt32))
        End If
    End Sub

    Private Sub XenonButton6_Click(sender As Object, e As EventArgs) Handles XenonButton6.Click
        If Proc IsNot Nothing AndAlso Not Proc.HasExited Then Proc.Kill()
    End Sub

    Private Sub XenonButton13_Click(sender As Object, e As EventArgs) Handles XenonButton13.Click
        If IO.File.Exists(XenonTextBox1.Text) AndAlso IO.Path.GetExtension(XenonTextBox1.Text).ToUpper = ".SCR" Then Shell("""" & XenonTextBox1.Text & """" & " /s")
    End Sub

    Private Sub XenonButton14_Click(sender As Object, e As EventArgs) Handles XenonButton14.Click
        If Proc IsNot Nothing AndAlso Not Proc.HasExited Then Proc.Kill()
        If IO.File.Exists(XenonTextBox1.Text) AndAlso IO.Path.GetExtension(XenonTextBox1.Text).ToUpper = ".SCR" Then
            Proc = Process.GetProcessById(Shell("""" & XenonTextBox1.Text & """" & " /c", AppWinStyle.NormalFocus))
            Proc.WaitForExit()
            Proc = Process.GetProcessById(Shell("""" & XenonTextBox1.Text & """" & " /p " & pnl_preview.Handle.ToInt32))
        End If
    End Sub

    Private Sub ScreenSaver_Editor_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        If Proc IsNot Nothing AndAlso Not Proc.HasExited Then Proc.Kill()
    End Sub

    Private Sub XenonTrackbar5_Scroll(sender As Object) Handles XenonTrackbar5.Scroll
        XenonButton4.Text = sender.Value
    End Sub

    Private Sub XenonButton4_Click(sender As Object, e As EventArgs) Handles XenonButton4.Click
        Dim response As String = InputBox(My.Lang.InputValue, sender.text, My.Lang.ItMustBeNumerical)
        sender.Text = Math.Max(Math.Min(Val(response), XenonTrackbar5.Maximum), XenonTrackbar5.Minimum) : XenonTrackbar5.Value = Val(sender.Text)
    End Sub

    Private Sub XenonButton259_Click(sender As Object, e As EventArgs) Handles XenonButton259.Click

        If OpenThemeDialog.ShowDialog = DialogResult.OK Then
            Using _Def As CP = CP_Defaults.From(My.PreviewStyle)
                GetFromClassicThemeFile(OpenThemeDialog.FileName, _Def.ScreenSaver)
            End Using
        End If
    End Sub

    Sub GetFromClassicThemeFile(File As String, _DefaultScrSvr As CP.Structures.ScreenSaver)
        Using _ini As New INI(File)
            XenonTextBox1.Text = _ini.IniReadValue("boot", "SCRNSAVE.EXE", _DefaultScrSvr.File).PhrasePath
        End Using
    End Sub

    Private Sub Form_HelpButtonClicked(sender As Object, e As CancelEventArgs) Handles Me.HelpButtonClicked
        Process.Start(My.Resources.Link_Wiki & "/Edit-Screen-Saver")
    End Sub
End Class

File: WinPaletter/Extras/ScreenSaver_Editor.cs
using Microsoft.VisualBasic;
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Windows.Forms;
using WinPaletter.Theme;

namespace WinPaletter
{

    public partial class ScreenSaver_Editor
    {

        private Process Proc;

        public ScreenSaver_Editor()
        {
            InitializeComponent();
        }

        private void ScreenSaver_Editor_Load(object sender, EventArgs e)
        {
            Button12.Image = Forms.MainFrm.Button20.Image.Resize(16, 16);
            pnl_preview.DoubleBuffer();
            ApplyFromTM(Program.TM);

            this.LoadLanguage();
            ApplyStyle(this);
        }

        public void ApplyFromTM(Theme.Manager TM)
        {
            ScrSvrEnabled.Checked = TM.ScreenSaver.Enabled;
            TextBox1.Text = TM.ScreenSaver.File;
            Trackbar5.Value = TM.ScreenSaver.TimeOut;
            CheckBox1.Checked = TM.ScreenSaver.IsSecure;
        }

        public void ApplyToTM(Theme.Manager TM)
        {
            TM.ScreenSaver.Enabled = ScrSvrEnabled.Checked;
            TM.ScreenSaver.File = TextBox1.Text;
            TM.ScreenSaver.TimeOut = Trackbar5.Value;
            TM.ScreenSaver.IsSecure = CheckBox1.Checked;
        }

        private void Button11_Click(object sender, EventArgs e)
        {
            if (OpenFileDialog1.ShowDialog() == DialogResult.OK)
            {
                Theme.Manager TMx = new(Theme.Manager.Source.File, OpenFileDialog1.FileName);
                ApplyFromTM(TMx);
                TMx.Dispose();
            }
        }

        private void Button9_Click(object sender, EventArgs e)
        {
            Theme.Manager TMx = new(Theme.Manager.Source.Registry);
            ApplyFromTM(TMx);
            TMx.Dispose();
        }

        private void Button12_Click(object sender, EventArgs e)
        {
            using (Manager _Def = Theme.Default.Get(Program.PreviewStyle))
            {
                ApplyFromTM(_Def);
            }
        }

        private void Button8_Click(object sender, EventArgs e)
        {
            ApplyToTM(Program.TM);
            Close();
        }

        private void Button10_Click(object sender, EventArgs e)
        {
            Cursor = Cursors.WaitCursor;
            Theme.Manager TMx = new(Theme.Manager.Source.Registry);
            ApplyToTM(TMx);
            ApplyToTM(Program.TM);
            TMx.ScreenSaver.Apply();
            TMx.Dispose();
            Cursor = Cursors.Default;
        }

        private void Button7_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void ScrSvrEnabled_CheckedChanged(object sender, EventArgs e)
        {
            checker_img.Image = ((UI.WP.Toggle)sender).Checked ? Properties.Resources.checker_enabled : Properties.Resources.checker_disabled;
        }

        private void Button5_Click(object sender, EventArgs e)
        {
            if (System.IO.File.Exists(TextBox1.Text) && System.IO.Path.GetExtension(TextBox1.Text).ToUpper() == ".SCR")
            {
                if (Proc is not null && !Proc.HasExited)
                    Proc.Kill();
                Proc = Process.GetProcessById(Interaction.Shell($"\"{TextBox1.Text}\" /p {pnl_preview.Handle.ToInt32()}"));
            }
        }


        private void Button1_Click(object sender, EventArgs e)
        {
            if (OpenFileDialog2.ShowDialog() == DialogResult.OK)
            {
                TextBox1.Text = OpenFileDialog2.FileName;
            }
        }

        private void TextBox1_TextChanged(object sender, EventArgs e)
        {
            if (System.IO.File.Exists(TextBox1.Text) && System.IO.Path.GetExtension(TextBox1.Text).ToUpper() == ".SCR")
            {
                if (Proc is not null && !Proc.HasExited)
                    Proc.Kill();
                Proc = Process.GetProcessById(Interaction.Shell($"\"{TextBox1.Text}\" /p {pnl_preview.Handle.ToInt32()}"));
            }
        }

        private void Button6_Click(object sender, EventArgs e)
        {
            if (Proc is not null && !Proc.HasExited)
                Proc.Kill();
        }

        private void Button13_Click(object sender, EventArgs e)
        {
            if (System.IO.File.Exists(TextBox1.Text) && System.IO.Path.GetExtension(TextBox1.Text).ToUpper() == ".SCR")
                Interaction.Shell($"\"{TextBox1.Text}\" /s");
        }

        private void Button14_Click(object sender, EventArgs e)
        {
            if (Proc is not null && !Proc.HasExited)
                Proc.Kill();
            if (System.IO.File.Exists(TextBox1.Text) && System.IO.Path.GetExtension(TextBox1.Text).ToUpper() == ".SCR")
            {
                Proc = Process.GetProcessById(Interaction.Shell($"\"{TextBox1.Text}\" /c", AppWinStyle.NormalFocus));
                Proc.WaitForExit();
                Proc = Process.GetProcessById(Interaction.Shell($"\"{TextBox1.Text}\" /p {pnl_preview.Handle.ToInt32()}"));
            }
        }

        private void ScreenSaver_Editor_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (Proc is not null && !Proc.HasExited)
                Proc.Kill();
        }

        private void Trackbar5_Scroll(object sender)
        {
            Button4.Text = ((UI.WP.Trackbar)sender).Value.ToString();
        }

        private void Button4_Click(object sender, EventArgs e)
        {
            string response = InputBox(Program.Lang.InputValue, ((UI.WP.Button)sender).Text, Program.Lang.ItMustBeNumerical);
            ((UI.WP.Button)sender).Text = Math.Max(Math.Min(Conversion.Val(response), Trackbar5.Maximum), Trackbar5.Minimum).ToString();
            Trackbar5.Value = (int)Math.Round(Conversion.Val(((UI.WP.Button)sender).Text));
        }

        private void Button259_Click(object sender, EventArgs e)
        {

            if (OpenThemeDialog.ShowDialog() == DialogResult.OK)
            {
                using (Manager _Def = Theme.Default.Get(Program.PreviewStyle))
                {
                    GetFromClassicThemeFile(OpenThemeDialog.FileName, _Def.ScreenSaver);
                }
            }
        }

        public void GetFromClassicThemeFile(string File, Theme.Structures.ScreenSaver _DefaultScrSvr)
        {
            using (INI _ini = new(File))
            {
                TextBox1.Text = _ini.Read("boot", "SCRNSAVE.EXE", _DefaultScrSvr.File).PhrasePath();
            }
        }

        private void Form_HelpButtonClicked(object sender, CancelEventArgs e)
        {
            Process.Start($"{Properties.Resources.Link_Wiki}/Edit-Screen-Saver");
        }
    }
}